api/azure/api.go

```diff
@@ -96,24 +99,37 @@ func (a apiImpl) FetchAzureArtifact(artifact BuildArtifact, w io.Writer) error {
 		return err
 	}
 
-	// Extract the report from the artifact.
-	reportPath := fmt.Sprintf("%s/wpt_report.json", artifact.Name)
 	data, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		log.Errorf("Failed to read response body")
 		return err
 	}
+
+	// Extract the report from the artifact.
+	return extractReports(a.ctx, artifact.Name, data, writer)
+}
+
+// extractReports extracts report files from the given zip.
+func extractReports(ctx context.Context, artifactName string, data []byte, writer *multipart.Writer) error {
+	log := shared.GetLogger(ctx)
+	reportPath, err := regexp.Compile(fmt.Sprintf(`%s/wpt_report.*\.json$`, artifactName))
+	if err != nil {
+		return err
+	}
+	extracted := 0
 	z, err := zip.NewReader(bytes.NewReader(data), int64(len(data)))
 	for _, f := range z.File {
-		if f.Name == reportPath {
+		if reportPath.MatchString(f.Name) {
+			fileName := f.Name[len(artifactName)+1:]
+			fileField, err := writer.CreateFormFile("result_file", fileName)
 			var fileData io.ReadCloser
 			if fileData, err = f.Open(); err != nil {
-				log.Errorf("Failed to extract %s", reportPath)
+				log.Errorf("Failed to extract %s", f.Name)
 				return err
 			}
 			defer fileData.Close()
```

Comment:
This is problematic. `defer` pushes the execution to the end of the **function**, but to the end of the current scope. Hence, we would hold onto a lot of readers.

I'd suggest wrapping everything inside this `if` block in a function, because the alternative (explicitly calling `Close()` on every exit path is more error-prone).

Reply:
Scoped.